### PR TITLE
Initial try for an auto-detach observer

### LINKED ISSUE #47: Dispose subscription when observable terminates
Currently a subscription to an observable is only disposed when the subscriber disposes it. If I remember correctly Rx.Net additionally disposes a subscription when the observable terminates (successfully or with an error).

Given the following code:

```fsharp
let o = AsyncRx.take 5 fileChangedObservable // implemented using `FileSystemWatcher`
let! sub = o.SubscribeAsync onFileChanged
```

The resources used for the `FileSystemWatcher` are not released until someone calls `sub.DisposeAsync ()` - even if all of the 5 events have been triggered.

I don't see an easy way to do this because there are so many places where anonymous objects that implement `IAsyncObservable` are created and I don't see an easy way to add the disposal of the subscription to all of them.
In the best case it's not even possible any more to create an `IAsyncObservable` that *doesn't* dispose the subscription on termination.
What do you think?

### PR DESCRIPTION
**Pull request by dbrattli**
Fixes #47

### CODE REVIEW

File: AsyncRx/src/AsyncObserver.fs [hunk 1]
```diff
@@ -86,3 +86,18 @@ module AsyncObserver =
             }
         }
 
+    let autoDetachObserver (obv: IAsyncObserver<'TSource>) : IAsyncObserver<'TSource>*(Async<IAsyncRxDisposable>->Async<IAsyncRxDisposable>) =
+        let disposables : ResizeArray<IAsyncRxDisposable> = ResizeArray ()
```
Comment (johannesegger) [inline AsyncRx/src/AsyncObserver.fs]: Is it a problem that `ResizeArray` is not thread-safe?

File: AsyncRx/src/AsyncObserver.fs [hunk 2]
```diff
@@ -86,3 +86,18 @@ module AsyncObserver =
...
+            for disp in disposables do
+                do! disp.DisposeAsync ()
+        }
+        let safeObv = AsyncDisposable.Create cancel |> safeObserver obv
+
+        // Auto-detaches (disposes) the disposable when the observer completes with success or error.
+        let autoDetach (disposable: Async<IAsyncRxDisposable>) = async {
+            let! disp = disposable
+            disposables.Add disp
+            return disp
```
Comment (johannesegger) [inline AsyncRx/src/AsyncObserver.fs:118]: Not sure, but shouldn't it return `AsyncDisposable.Create cancel`?

Comment (dbrattli) [reply AsyncRx/src/AsyncObserver.fs:118]: The auto-detach observer may dispose multiple disposables when it closes, but those individual disposables should not dispose each other. So it's just a pass-through, `id` function with a side-effect. Then you can do `obv |> source.SubscribeAsync |> autoDetach`

Comment (johannesegger) [reply AsyncRx/src/AsyncObserver.fs:118]: I see, you're right, thanks.

File: AsyncRx/src/AsyncObserver.fs (hunk 1 above)
Comment (dbrattli) [reply AsyncRx/src/AsyncObserver.fs]: Yes, replaced with a mailbox agent.

Comment (johannesegger) [reply AsyncRx/src/AsyncObserver.fs]: Looks good to me. Nice work.

Comment (dbrattli) [reply AsyncRx/src/AsyncObserver.fs]: Thanks, let's try this

**Pull request merged**